authors/apps/articles/models.py

```diff
@@ -129,3 +131,67 @@ class Favorite(models.Model):
         related_name='favorites')
     article_id = models.ForeignKey(
         Article, on_delete=models.CASCADE, related_name='favorites')
+
+
+class Tag(models.Model):
+    tag = models.CharField(max_length=100, db_index=True, unique=True)
+    slug = models.SlugField(max_length=240, db_index=True, unique=True)
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = slugify(self.tag)
+        super().save(*args, **kwargs)
+
+    def __str__(self):
+        return self.slug
+
+    objects = models.Manager()
+
+    def _create_tag(self, tag_name):
+        '''
+        This method checks if the tag exists and returns the tag object
+        else it creates a new tag object and returns it.
+        '''
+        found = Tag.objects.filter(slug=slugify(tag_name))
+        if len(found) < 1:
+            Tag.objects.create(tag=tag_name)
+        return Tag.objects.filter(slug=slugify(tag_name)).first()
+
+    def _remove_tags_without_articles(self, tag_name):
+        '''This method takes a tag name and checks if the tag has
+        associated articles and if not it deletes the tag from the
+        database in order to save on database space.
+        '''
+        found = Tag.objects.filter(slug=slugify(tag_name))
+        if found and len(found) > 0:
+            the_tag = Tag.objects.filter(slug=slugify(tag_name)).first()
+            its_articles = the_tag.articles.all()
+            if len(its_articles) < 1:
+                the_tag.delete()
+                return True
+        return False
+
+    def _update_article_tags(self, instance, new_tag_list):
+        '''This method is used to update the tags of an
+        article and do general tag mainenance.'''
+        old_tags = instance.tags.all()
+        old_tag_slugs = [x.slug for x in old_tags]
+        new_tag_slugs = [slugify(x) for x in new_tag_list]
+        # Check for items in the old list that are not in the new
+        # list and remove them.
+        tags_to_remove = list(set(old_tag_slugs) - set(new_tag_slugs))
+        remove_tag_objects = [x for x in old_tags if (x.slug in tags_to_remove)]
```

Comment:
@Etomovich, it is better to use a more descriptive variable name other than x in the Ternary Operators.